testing/architecture-test/src/test/java/org/gradle/architecture/test/PlatformBoundariesTest.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.architecture.test;
+
+import com.google.gson.Gson;
+import com.tngtech.archunit.base.DescribedPredicate;
+import com.tngtech.archunit.core.domain.JavaAccess;
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.junit.AnalyzeClasses;
+import com.tngtech.archunit.junit.ArchTest;
+import com.tngtech.archunit.lang.ArchCondition;
+import com.tngtech.archunit.lang.ArchRule;
+import com.tngtech.archunit.lang.ConditionEvents;
+import com.tngtech.archunit.lang.SimpleConditionEvent;
+import org.jspecify.annotations.Nullable;
+
+import java.io.FileReader;
+import java.io.Reader;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.CodeSource;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+import static org.gradle.architecture.test.ArchUnitFixture.freeze;
+import static org.gradle.architecture.test.ArchUnitFixture.safeReflect;
+
+@SuppressWarnings("unchecked")
+@AnalyzeClasses(packages = "org.gradle")
+public class PlatformBoundariesTest {
+
+    private static final List<Path> allPlatformDirs;
+    private static final Map<String, List<Path>> dirsByPlatform;
+    private static final Map<String, List<Path>> usedDirsByPlatform;
+    private static final Map<String, List<Path>> allowedDirsByPlatform;
+
+    static {
+        Path basePath = Paths.get(System.getProperty("org.gradle.architecture.platforms-base-path"));
+        Path jsonPath = Paths.get(System.getProperty("org.gradle.architecture.platforms-json"));
+        Map<String, Map<String, List<String>>> data;
+        try (Reader reader = new FileReader(jsonPath.toFile())) {
+            data = (Map<String, Map<String, List<String>>>) new Gson().fromJson(reader, Map.class);
```

Comment:
💅 Maybe this can be better expressed as an actual class (several classes) to give the user a glimpse of how json is structured. The stream dance below isn't easy to follow.

Reply:
A single class was enough